### ISSUE TITLE
[BUG] Bicep types changes the order of resources and modules.

### ISSUE THREAD

**Issue by MadsHT** [Azure/PSRule.Rules.Azure#3257]
### Description of the issue

I have some bicep code the deployes a network, subnets and related network security groups and just started looking into working ps-rule to check my code.
I have noticed that I failed the rule `Azure.VNET.UseNSGs`. But when I deployed the template all the subnets got network security groups as expected.

I make the NSGs in a module using a loop:
```bicep 
@batchSize(1)
module newNSGs './module-1.bicep' = [
  for snet in vnet.subnets: {
    name: 'nsg-${snet}-deployment'
    params: {
      name: 'nsg-${snet}'
    }
  }
]
```

And using the resulting module array in the virtual network when making the subnets:
```bicep 
resource virtualNetwork 'Microsoft.Network/virtualNetworks@2024-05-01' = {
  name: 'name'
  location: resourceGroup().location
  properties: {
    ...
    subnets: [
      for (snet, idx) in vnet.subnets: {
        name: snet
        properties: {
          addressPrefix: '10.0.${16*idx}.0/24'
          networkSecurityGroup: {
            id: newNSGs[idx].outputs.id
          }
        }
      }
    ]
  }
}
```

So i looked into the template the ps-rule generates using `export-AzRuleTemplateData` and noticed that the `networkSecurityGroup` object in the subnet was empty for all subnet but the last one.

I then tried to isolate the problem, and lastly i tried to remove the virtual network type the I was importing and changed the parameter to an object and then the problem when away.

 

### Reproduction

First make the following ps-rule.yaml:
```yaml
# ps-rule.yaml

include:
  # Use rules from the following modules
  module:
    - PSRule.Rules.Azure

configuration:
  # Set the Bicep parameter expansion configuration option to enable expansion of Azure Bicep parameter files (.bicepparam).
  AZURE_PARAMETER_FILE_EXPANSION: true
  # Enable expansion for Bicep source files.
  AZURE_BICEP_FILE_EXPANSION: true
  # Validate that the Bicep CLI is used
  AZURE_BICEP_CHECK_TOOL: true
  AZURE_BICEP_FILE_EXPANSION_TIMEOUT: 30
  AZURE_BICEP_MINIMUM_VERSION: "0.32.4"

# Require a minimum version of modules that include referenced baseline.
requires:
  PSRule: ">=2.9.0"
  PSRule.Rules.Azure: ">=1.40.0"
```

Then make two bicep files in the same folder. One will function as the module and one will be the main deployment file.

In the module file:
```bicep
// module.bicep

param name string

resource networkSecurityGroup 'Microsoft.Network/networkSecurityGroups@2019-11-01' = {
  name: name
  location: resourceGroup().location
  properties: {
    securityRules: [
      {
        name: 'nsgRule'
        properties: {
          description: 'description'
          protocol: 'Tcp'
          sourcePortRange: '*'
          destinationPortRange: '*'
          sourceAddressPrefix: '*'
          destinationAddressPrefix: '*'
          access: 'Allow'
          priority: 100
          direction: 'Inbound'
        }
      }
    ]
  }
}

output id string = networkSecurityGroup.id
```

in the main.bicep:
```bicep
// main.bicep

param vnet object = {
  name: 'hello'
  subnets: [
    'subnet1'
    'subnet2'
  ]
}

@batchSize(1)
module newNSGs './module.bicep' = [
  for snet in vnet.subnets: {
    name: 'nsg-${snet}-deployment'
    params: {
      name: 'nsg-${snet}'
    }
  }
]

resource virtualNetwork 'Microsoft.Network/virtualNetworks@2024-05-01' = {
  name: 'virtualNetwork'
  location: resourceGroup().location
  properties: {
    addressSpace: {
      addressPrefixes: [
        '10.0.0.0/16'
      ]
    }
    subnets: [
      for (snet, idx) in vnet.subnets: {
        name: snet
        properties: {
          addressPrefix: '10.0.${16*idx}.0/24'
          networkSecurityGroup: {
            id: newNSGs[idx].outputs.id
          }
        }
      }
    ]
  }
}
```

Running `invoke-PSRULE -InputPath main.bicep -outcome fail` should show one failed rule for the virtual network, which is `Azure.Resource.UseTags`.

Now add `type foo = string` to the top of the `main.bicep`. Running `invoke-PSRULE -InputPath main.bicep -outcome fail` should now show another rule failed for the virtual network `Azure.VNET.UseNSGs`

### Version of PSRule

2.9.0

### Version of PSRule for Azure

1.40.0

### Additional context

I tried to reproduce the same results without modules, changing the NSGs into a resource, but that seems to work as expected, even with the type pressent. So the combination of types and looped(?) modules seems to cause problems in the way the template is being generated.

**Comment by BernieWhite** [comment]
Thanks for the detailed examples in the issue @MadsHT, I'll investigate it.

In the meantime, try release v1.41.2 if you haven't already since some cross module issues have been fixed.

**Comment by MadsHT** [comment]
Thank you for the quick response, I tried updating to version 1.41.2, sadly it did not resolve the problem.

**Comment by BernieWhite** [comment]
This should be resolved in v1.41.3.

**Comment by MadsHT** [comment]
Awesome this seems to work as expected now :+1:
Thank you very much for the quick response and great work. :)
